### ISSUE TITLE
[New Feature] Plugin system capability in Flutter apps

### ISSUE THREAD

**Issue by JulienElkaim** [flutter/flutter#89201]
## Context

Modern applications often offer the possibility to extend their features thanks to a plugin system. It's a great opportunity to make flexible applications and unlock the user experience as they don't have to wait the next release to get new features.

## My problem  

I want to create such application using Dart & Flutter. This framework is a great choice as I want my app to be Desktop (Windows, OSX) & Mobile (Android, iOS), but its strength (AOT compilation) is also the problem here.  

To create a plugin system in my app, I need to provide the plugin developers with interfaces they will implement to extend my application features. Once a user wants to download and activate a plugin, I want to load the plugin's implementations of my interfaces in my runtime and be able to triger them. I want to be able to load/unload these implementations as the user enable/disable the plugin.

## Existing (but not convincing) solutions  

I did not find so much solutions. I ended up with 2, that I don't like:  

- Use a script language interpreter. I found examples as [Flutter Lua](https://github.com/drydart/flutter_lua), which embeds Lua interpreter to be able to execute Lua scripts. Its very annoying as it increases the app size a lot, for a poor integration (Looks like a simple exec at the end, so no interfaces contract possible here). Also it is definitely not dev-friendly : my project would be open source, I want developers contributing to the application core in Dart to use the same language when they want to create plugins.  

- Use webviews. Its definitely a lightweight solution, but I also feel it very unsafe, I am afraid of the possibilities offered to the malicious developers to redirect my users to scams / unsafe zones.  Also, it is here again an issue as I don't know how I would integrate the notion of interfaces with this solution...  

## Proposal  

I feel I am not knowledgeable enough in Dart/Flutter to propose a tangible solution, but here is a naive proposal : 

Maybe an embedded Dart interpreter can do the job ? Once the plugin needs to be loaded It may runs AOT compilation for the current platform, and make the new Objects and Widgets available for the application runtime. 
If so, I think it also requires a "sniffing" capability into flutter to be able to discover and use the new features available and integrate them to the application system.

I hope I was clear enough, don't hesitate if any questions !

**Comment by EmreSURK** [comment]
Hi,
Adding new code to the published app would be nice. Some web-based technologies have that but Apple does not accept any change on the app after it is published.

**Comment by JulienElkaim** [comment]
> Hi,
Adding new code to the published app would be nice. Some web-based technologies have that but Apple does not accept any change on the app after it is published.

Sadly.... you are right !
Btw, If my request is impossible. Talking about web based technology I think about a possible work around :

- Split Back & Front, Flutter & Dart is only doing the FrontEnd calling the Backend's API.
- Use a backend technology with Plugin capability (i.e, Java).
- When I need custom displays for Plugins: 
  - Backend provides FrontEnd with a JSON description (format to use, to be defined)
  - Frontend Flutter translates JSON into Widgets.  

Therefore, only Desktop version would be possible (Backend server run on same device), Mobile version has to connect to an instance of the backend API.

Not ideal, as we have to find a JSON format providing enough information to have a "smart" display....

**Comment by hpoul** [comment]
> Apple does not accept any change on the app after it is published

Luckily Apple controls only 2 out of 7(?) platforms. And flutter aims to <[avoid the “lowest common denominator” problem](https://flutter.dev/docs/resources/faq#can-i-access-platform-services-and-apis-like-sensors-and-local-storage)> 😅️

btw, this might be related to: 

* #14330

(although I guess the way to "extend" android apps would be through content providers or providing intent filters?)

**Comment by JulienElkaim** [comment]
Thank you for the link, hpoul! 

You are right; this is the same issue we are talking about (having a JIT compilation). However, this comment was in 2019; maybe after Flutter 2 and the shipping of a stable Flutter Desktop version, the Flutter Team can reconsider their position?..
It is more a "Desktop apps feature" rather than a Mobile App one.

The comment below mentions a possible work-around to my problem with making a "Declarative Programmation" of widgets (package RFW). It ships a widget into a "RemoteWidget", already present in the running application. 

https://github.com/flutter/flutter/issues/14330#issuecomment-921203049

I may give it a try, but the lack of JIT compilation for a more advanced plugin system is a big problem in 2021, so I am considering using a more "mature" technology for Desktops apps, like Electron.

**Comment by chgibb** [comment]
@JulienElkaim have you looked at https://hydro-sdk.io/ ?

**Comment by SzczurekYT** [comment]
Yes please, now when flutter has full support for desktop, that is way more advanced then mobile phones this is must have. Extensions are everywhere. In out IDE's, browsers, note apps, mail clients, games (mods). Because they are a great concept.
Unfortunetly i think that they don't care too much about this.
This is open for **10 years!** 
https://github.com/dart-lang/sdk/issues/10530

**Comment by JulienElkaim** [comment]
I did not really follow what happened in Flutter the past year, but yes even at this time it seemed not their priority. Bad decision, as you said plugins are a must to have on a desktop platform

**Comment by alexrintt** [comment]
In case of Android platform, we could let the plugins be standalone apks and call it from Flutter app using intents (note: https://pub.dev/packages/android_intent_plus does not support intent results, so you will need to implement your own Flutter intent plugin).

But in this case, I've to say, it's much more easier to write a native android app that we can simply use intents without needing to create platform plugins.

And in case of a desktop app, Electron, I don't like cause it seems just a chromium instance, but that's the way, JavaScript is the JIT we need for runtime plugins (approach using Electron: https://stackoverflow.com/questions/52296921/how-to-implement-plugins-into-electron-app).

**Comment by SzczurekYT** [comment]
> I did not really follow what happened in Flutter the past year, but yes even at this time it seemed not their priority. Bad decision, as you said plugins are a must to have on a desktop platform

I think they worry too much about mobile / don't care enough about desktop. They say that stores do not accept code loading, but on desktop stores are used very rarely, and desktop is half of the supported platforms.
The desktop support landed some time ago, but I feel it's kinda incomplete with some features missing or you needing to use a package for some basic stuff.

**Comment by yjbanov** [comment]
Throwing out an idea: there's experimental support for compiling Dart to WebAssembly for runtimes that support the WasmGC spec. Wasm provides a tight sandbox for executing untrusted code, such as plugin code written by a third party for your app. It would be interesting to see if it's possible to design a plugin system where the plugin code is written in Dart, compiled to wasm, and loaded by the app as a wasm module. Wasm's import/export system can be used for the app to define a safe API surface for the plugin to interact with the app. It should be possible to try in Chrome and Edge today, and if it works well, can be extended to other platforms in the future by embedding a wasm VM, such as WasmTime, when those runtimes catch up with the WasmGC spec.

**Comment by iapicca** [comment]
> if it's possible to design a plugin system where the plugin code is written in Dart, compiled to wasm

@yjbanov This would be cool, 
but I got a cold shower from this [comment](https://github.com/dart-lang/sdk/issues/32894#issuecomment-1564609414).

For my understanding, 
seems  that `dart2wasm` is intended to be a "web only" (or even chrome only) feature 
for the short-middle term.

This is a huge shame, not just for flutter itself, for my dart backend fetish and [more](https://github.com/dart-lang/sdk/issues/32894#issuecomment-1564912832)

I really hope I'm wrong

**Comment by RedyAu** [comment]
The packages https://pub.dev/packages/dart_eval and https://pub.dev/packages/flutter_eval seem promising in this regard.

**Comment by Shabinder** [comment]
Would love to see this being implemented, WASM is a great option to have since having that VM in. any platform like Android / iOS wont break store rules as far as I am aware, being the same logic as being used in here: https://github.com/cashapp/zipline

**Comment by EchoEllet** [comment]
> Modern applications often offer the possibility to extend their features thanks to a plugin system. It's a great opportunity to make flexible applications and unlock the user experience as they don't have to wait the next release to get new features.

You can use packaging & self-updating solutions for desktop applications like [Hydraulic Conveyor](https://conveyor.hydraulic.dev/14.3/tutorial/hare/flutter/), it can be used for JVM (Java FX, Swing, or Compose Multiplatform), Electron or Native targets:

```
Build self-updating desktop app packages in minutes. Deploy to every OS from any OS. Platform native formats, fully signed and notarized.
```

It's pretty customizable, and supports delta updates, making the process faster and the signing process less expensive.

It requires payment for commercial apps, though it is not required for open-source projects.

You can take a look at other alternatives too that have support for Flutter.

> I did not really follow what happened in Flutter the past year, but yes even at this time it seemed not their priority. Bad decision, as you said plugins are a must to have on a desktop platform

There are several reasons why Flutter and Dart have limited support for executing code dynamically in a direct way.

Flutter is primarily designed as a cross-platform UI framework and has evolved beyond that.

JVM and similar platforms support features like **Dynamic Class Loading**, **Reflections** (**Java Swing** depends on it internally), and calling functions from JAR files, they lack features such as **Hot Reload**, **Hot Restart**, **Shrinking**, and **Obfuscation** that Flutter supports. These dynamic features can complicate the implementation of other features or sometimes make it impossible to implement. Flutter's design choices aim to balance these needs for a stable and efficient development experience.

It's easier to implement in JS as it's dynamically typed, and most desktop applications with it usually use an embedded browser, either embedding Chromium or using the System WebView.

Usually, this feature is a requirement if you're developing an advanced project, such as an IDE, Web Browser, Game engine, 3D modeling, or WordPress.

Generally, a plugin system is not a standard feature in native desktop or mobile applications.

**Comment by mcquenji** [comment]
I'm currently working on a [plugin system](https://github.com/mcquenji/dscript) for flutter/dart if anyone is interested or wants to help :)

**Comment by CGSiemensEnergy** [comment]
Would be nice to have such functionality. Also, dynamic modules would be an option as well. :)
